File: src/Functional/OCaml/example1.ml
module type UIFactory = sig
  val create_button : unit -> unit
  val create_checkbox : unit -> unit
end

module DarkFactory : UIFactory = struct
  let create_button () = print_endline "Dark Button"
  let create_checkbox () = print_endline "Dark Checkbox"
end

module LightFactory : UIFactory = struct
  let create_button () = print_endline "Light Button"
  let create_checkbox () = print_endline "Light Checkbox"
end

let create_ui_components (factory : (module UIFactory)) =
  let module F = (val factory : UIFactory) in
  F.create_button ();
  F.create_checkbox ()

(* Usage *)
let () =
  create_ui_components (module DarkFactory);
  create_ui_components (module LightFactory)

File: src/Functional/Erlang/example1.erl
-module(example1).
-export([dark_factory/0, light_factory/0, create_ui_components/1, main/0]).

%% A factory is one value that owns the constructors for an entire family.
dark_factory() ->
    #{button => fun dark_button/0, checkbox => fun dark_checkbox/0}.

light_factory() ->
    #{button => fun light_button/0, checkbox => fun light_checkbox/0}.

create_ui_components(Factory) ->
    CreateButton = maps:get(button, Factory),
    CreateCheckbox = maps:get(checkbox, Factory),
    Button = CreateButton(),
    Checkbox = CreateCheckbox(),
    io:format("~s~n~s~n", [Button, Checkbox]).

main() ->
    create_ui_components(dark_factory()),
    create_ui_components(light_factory()).

%% Concrete products stay private; clients receive them through one family value.
dark_button() -> "Dark Button".
light_button() -> "Light Button".
dark_checkbox() -> "Dark Checkbox".
light_checkbox() -> "Light Checkbox".
